stlab/future.hpp

```diff
@@ -564,59 +572,59 @@ class future<T, detail::enable_if_copyable<T>> {
 
     future() = default;
 
-    bool valid() const { return static_cast<bool>(_p); }
+    bool valid() const { return static_cast<bool>(std::atomic_load(&_p)); }
```

Comment:
A future should not be shared by multiple threads (there is a shared resource linking futures but you copy the future to a thread). So this atomic_load doesn't do anything but slow us down.

cancel_try() is trying to reset() the shared_ptr but only if no other future owns it. But here are the problems:

`if (!_p.unique()) return false; 

`unique()` is deprecated, second `use_count()` What does a use_count() of 1 tell us? That this is the only future holding the shared state, and at the moment, the shared state isn't being set (because then the weak_ptr would aquire a lock). Okay...

`std::weak_ptr<> p = _p`

We want to see if it really goes away...

_p.reset(); 

Let go of our shared ptr - note at this case it could be in the middle of setting the shared state so a weak ptr could have a lock.

_p = p.lock();

If by chance, we catch it at the exact moment it was being set and our reset() and lock() happen during that instant. Then we get it back. Very unlikely. But if that is the case, then it is now ready and so we don't cancel?

So I don't see how this can be any use. I don't know what I was thinking except that cancel should maybe be symmetric to get_try? I think I was sleep deprived.

I say we replace cancel_try() with reset() that looks like:

`void reset() { _p.reset(); }`

Which is pretty much what I wrote in the comment - except I don't think we should call it cancel() - because it doesn't. It simply says that we no longer care about the result of this future. We don't need this because it is the same as assigning a default constructed value, but that requires <>s.

So dump all the atomic_load, nuke cancle_try(), and add reset(). And we'll also close issue #17. Good?